tezos/new_context/src/working_tree/storage.rs

```diff
@@ -35,76 +60,131 @@ impl Default for TreeStorageId {
 }
 
 impl TreeStorageId {
-    fn try_new_tree(start: usize, end: usize) -> Result<Self, StorageIdError> {
+    fn try_new_tree(start: usize, end: usize) -> Result<Self, StorageError> {
         let length = end
             .checked_sub(start)
-            .ok_or(StorageIdError::TreeInvalidStartEnd)?;
+            .ok_or(StorageError::TreeInvalidStartEnd)?;
 
-        if start & !0x3FFFFFFF != 0 {
-            // Must fit in 30 bits
-            return Err(StorageIdError::TreeStartTooBig);
+        if start & !0xFFFFFFFF != 0 {
+            // Must fit in 32 bits
+            return Err(StorageError::TreeStartTooBig);
         }
 
-        if length & !0xFFFFF != 0 {
-            // Must fit in 20 bits
-            return Err(StorageIdError::TreeLengthTooBig);
+        if length & !0xFFFFFFF != 0 {
+            // Must fit in 28 bits
+            return Err(StorageError::TreeLengthTooBig);
         }
 
-        let tree_id = Self::new()
-            .with_start(start as u32)
-            .with_length(length as u32);
+        let tree_id = Self {
+            bits: (start as u64) << 28 | length as u64,
+        };
 
         debug_assert_eq!(tree_id.get(), (start as usize, end));
 
         Ok(tree_id)
     }
 
+    fn try_new_inode(index: usize) -> Result<Self, StorageError> {
+        if index & !0xFFFFFFFFFFFFFFF != 0 {
+            // Must fit in 60 bits
+            return Err(StorageError::InodeIndexTooBig);
+        }
+
+        Ok(Self {
+            bits: 1 << 60 | index as u64,
+        })
+    }
+
+    pub fn is_inode(&self) -> bool {
+        self.bits >> 60 != 0
+    }
+
+    pub fn get_inode_id(self) -> Option<InodeId> {
+        if self.is_inode() {
+            Some(InodeId(self.get_inode_index() as u32))
+        } else {
+            None
+        }
+    }
+
     fn get(self) -> (usize, usize) {
-        let start = self.start() as usize;
-        let length = self.length() as usize;
+        debug_assert!(!self.is_inode());
+
+        let start = (self.bits as usize) >> 28;
+        let length = (self.bits as usize) & 0xFFFFFFF;
 
         (start, start + length)
     }
 
+    /// Return the length of the small tree.
+    ///
+    /// Use `Storage::tree_len` to get the length of all trees (including inodes)
+    fn small_tree_len(self) -> usize {
+        debug_assert!(!self.is_inode());
+
+        (self.bits as usize) & 0xFFFFFFF
```

Comment:
@sebastiencs let's give these bitmasks names, to make this easier to understand.